### PR TITLE
[Example] Rgcn link-prediction using neighbor sampling.

### PR DESCRIPTION
**Pull request by classicsong**
## Description
#1629 #1695 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/rgcn/model.py [hunk 1]
```diff
@@ -97,7 +98,11 @@ def __init__(self,
                 self.embeds[str(ntype)] = embed
 
         self.node_embeds = th.nn.Embedding(node_tids.shape[0], self.embed_size, sparse=self.sparse_emb)
-        nn.init.uniform_(self.node_embeds.weight, -1.0, 1.0)
+        if gamma <= 0:
+            nn.init.uniform_(self.node_embeds.weight, -1.0, 1.0)
+        else:
+            emb_init = gamma / embed_size
+            nn.init.uniform_(self.node_embeds.weight, -emb_init, emb_init)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/model.py:105]: What is this initialization?

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 2]
```diff
@@ -0,0 +1,1147 @@
...
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.multiprocessing as mp
+from torch.multiprocessing import Queue
+from torch.nn.parallel import DistributedDataParallel
+from torch.utils.data import DataLoader
+import dgl
+import dgl.function as fn
+from dgl import DGLGraph
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:21]: delete

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 3]
```diff
@@ -0,0 +1,1147 @@
...
+
+    score = th.cat([pos_score, h_neg_score, t_neg_score])
+    label = th.cat([th.full((pos_score.shape), 1.0),
+                    th.full((h_neg_score.shape), 0.0),
+                    th.full((t_neg_score.shape), 0.0)]).to(score.device)
+    predict_loss = F.binary_cross_entropy_with_logits(score, label)
+    #pos_score = pos_score.mean()
+    #h_neg_score = h_neg_score.mean()
+    #t_neg_score = t_neg_score.mean()
+    #predict_loss = -(2 * pos_score + h_neg_score + t_neg_score) / 2
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:255]: Are these comment-out lines useful?

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 4]
```diff
@@ -0,0 +1,1147 @@
...
+        test_g.edata['norm'] = th.full((eid.shape[0],1), 1.)
+        for rel_id in range(edge_rels):
+            idx = (test_g.edata['etype'] == rel_id)
+            u_r = u[idx]
+            v_r = v[idx]
+            _, inverse_index, count = th.unique(v_r, return_inverse=True, return_counts=True)
+            degrees = count[inverse_index]
+            norm = th.ones(v_r.shape[0]) / degrees
+            norm = norm.unsqueeze(1)
+            test_g.edata['norm'][idx] = norm
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:1023]: Too many duplicate codes. Put them in a function.

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 5]
```diff
@@ -0,0 +1,1147 @@
...
+    test_g.edata['norm'].share_memory_()
+    test_g.edata['etype'].share_memory_()
+    test_g.ndata['ntype'].share_memory_()
+
+    train_g.create_formats_()
+    valid_g.create_formats_()
+    test_g.create_formats_()
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:1030]: comments about why need share_memory_ and create_formats_

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 6]
```diff
@@ -0,0 +1,1147 @@
...
+                        layer_norm=args.layer_norm)
+    if dev_id >= 0:
+        model.cuda(dev_id)
+        if args.mix_cpu_gpu is False:
+            embed_layer.cuda(dev_id)
+            embed_layer.node_embeds.cuda(dev_id)
+
+    if n_gpus > 1:
+        embed_layer = DistributedDataParallel(embed_layer, device_ids=[dev_id], output_device=dev_id)
+        #model = DistributedDataParallel(model, device_ids=[dev_id], output_device=dev_id, find_unused_parameters=True)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:747]: remove

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 7]
```diff
@@ -0,0 +1,1147 @@
...
+            if args.sampler == 'neighbor':
+                if queue is None:
+                    reg_loss = regularization_loss(th.cat([p_feats, n_feats]), model.w_relation)
+                else:
+                    reg_loss = regularization_loss(th.cat([p_feats, n_feats]), model.module.w_relation)
+            else:
+                if queue is None:
+                    reg_loss = regularization_loss(mb_feats, model.w_relation)
+                else:
+                    reg_loss = regularization_loss(mb_feats, model.module.w_relation)
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:821]: If the "neighbor" and "path" -based training are so different (saw so many branches), I'd suggest just have a `run_neighbor` and `run_path`.

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 8]
```diff
@@ -0,0 +1,1147 @@
...
+        pseeds = th.tensor(seeds).long()
+        pseeds = pseeds.squeeze()
+        bsize = pseeds.shape[0]
+
+        g = self.g
+        subg = g.edge_subgraph(pseeds)
+        p_u, p_v = subg.edges()
+        p_rel = subg.edata['etype']
+        org_nid = subg.ndata[dgl.NID]
+        #subg.ndata['type_id'] = g.ndata['type_id'][subg.ndata[dgl.NID]]
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:309]: remove

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 9]
```diff
@@ -0,0 +1,1147 @@
...
+                n_frontier = dgl.in_subgraph(g, n_curr)
+            else:
+                p_frontier = dgl.sampling.sample_neighbors(g, p_curr, fanout)
+                n_frontier = dgl.sampling.sample_neighbors(g, n_curr, fanout)
+            p_block = dgl.to_block(p_frontier, p_curr)
+            n_block = dgl.to_block(n_frontier, n_curr)
+            p_curr = p_block.srcdata[dgl.NID]
+            n_curr = n_block.srcdata[dgl.NID]
+            p_blocks.insert(0, p_block.int())
+            n_blocks.insert(0, n_block.int())
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:383]: there is no removal of the predicted links?

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 10]
```diff
@@ -0,0 +1,1147 @@
...
+            blocks.insert(0, block.int())
+
+        return seeds, blocks
+
+class LinkPathSampler:
+    """ Path based sampler.
+
+    In Path sampler, we randomly select K positive edges (e.g., 30000) to construct
+    a subgraph and add corresponding reverse edges. Then we randomly remove half of
+    positive edges in order to over information leakage.
```
Comment (jermainewang) [inline examples/pytorch/rgcn/link_predict_mp.py:288]: Is this a standard approach? Look weird to me esp. the removing half of the positive edges part.

"in order to over ..." <--------- miss a word?

File: examples/pytorch/rgcn/model.py (hunk 1 above)
Comment (classicsong) [reply examples/pytorch/rgcn/model.py:105]: I borrow this initialization from KE, this is for DistMult model, RGCN uses the same score function.

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 11]
```diff
@@ -0,0 +1,1147 @@
...
+        """
+        Forward with the RGCN model.
+        If the `path` sample is used, the input p_blocks is a subgraph and a full-graph forward is
+        applied for each layer.
+        If the `neighbor` sample is used, the p_blocks stores blocks to generate embeddings for
+        positive pairs and n_blocks stores blocks to generate embeddigns for negative pairs.
+
+        Parameters
+        ----------
+        p_blocks: DGLGraph or list of DGLGraph
```
[the remark this answers is not in the document]
Comment (classicsong) [reply examples/pytorch/rgcn/link_predict_mp.py:118]: Yes

File: examples/pytorch/rgcn/link_predict_mp.py [hunk 12]
```diff
@@ -0,0 +1,1147 @@
...
+                self.num_bases, activation=act, self_loop=self.use_self_loop,
+                low_mem=low_mem, dropout=self.dropout, layer_norm=layer_norm))
+
+    def forward(self, p_blocks, p_feats, n_blocks=None, n_feats=None, sample='neighbor'):
+        """
+        Forward with the RGCN model.
+        If the `path` sample is used, the input p_blocks is a subgraph and a full-graph forward is
+        applied for each layer.
+        If the `neighbor` sample is used, the p_blocks stores blocks to generate embeddings for
+        positive pairs and n_blocks stores blocks to generate embeddigns for negative pairs.
```
[the remark this answers is not in the document]
Comment (classicsong) [reply examples/pytorch/rgcn/link_predict_mp.py:114]: path is from original implementation
neighbor is just neighbor sampling.

File: examples/pytorch/rgcn/link_predict_mp.py (hunk 3 above)
Comment (classicsong) [reply examples/pytorch/rgcn/link_predict_mp.py:255]: Yes, different ways of achieving the loss.
Maybe we can remove it.

File: examples/pytorch/rgcn/link_predict_mp.py (hunk 9 above)
Comment (classicsong) [reply examples/pytorch/rgcn/link_predict_mp.py:383]: Yes, otherwise, it will be very slow.
